app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt

```diff
@@ -5,106 +5,197 @@ import android.graphics.Bitmap
 import android.graphics.Canvas
 import android.graphics.Matrix
 import android.graphics.Paint
-import android.graphics.Path
-import android.graphics.PorterDuff
-import android.graphics.PorterDuffXfermode
+import android.os.Parcel
+import android.os.Parcelable
+import android.view.ViewConfiguration
+import androidx.core.graphics.withMatrix
 import net.aiscope.gdd_app.R
+import java.util.*
+import kotlin.math.abs
 
-const val TWENTY = 20.0f
-
+@Suppress("TooManyFunctions")
 class MaskLayer(
     private val context: Context,
     private val scaleMatrix: Matrix
 ) {
+    companion object {
+        private const val TWENTY = 20.0f
+        private const val MATRIX_SIZE = 9
+        private const val MASK_PAINT_ALPHA = 200
 
-    private var viewHeight: Int = 0
-    private var viewWidth: Int = 0
-    private var maskBitmap: Bitmap? = null
-    private var maskCanvas: Canvas? = null
-    private val maskPaintBitmap = Paint().apply {
-        alpha = MASK_PAINT_ALPHA
-    }
-    private val maskPaintBrush = Paint().apply {
-        color = context.getColor(R.color.colorPrimary)
-        isAntiAlias = true
-        isDither = true
-        style = Paint.Style.STROKE
-        strokeJoin = Paint.Join.ROUND
-        strokeCap = Paint.Cap.ROUND
-    }
-    private val maskPaintErase = Paint().apply {
-        xfermode = PorterDuffXfermode(PorterDuff.Mode.CLEAR)
-        isAntiAlias = true
-        isDither = true
-        style = Paint.Style.STROKE
-        strokeJoin = Paint.Join.ROUND
-        strokeCap = Paint.Cap.ROUND
+        private fun newDefaultPaintBrush(color: Int) = Paint().apply {
+            this.isAntiAlias = true
+            this.isDither = true
+            this.style = Paint.Style.STROKE
+            this.strokeJoin = Paint.Join.ROUND
+            this.strokeCap = Paint.Cap.ROUND
+            this.color = color
+            this.alpha = MASK_PAINT_ALPHA
+        }
     }
-    private var maskPaint = maskPaintBrush
+
+    private var pathsAndPaints: MutableList<Pair<PointToPointPath, Paint>> = LinkedList()
+    private var undoPendingPaths = 0
+    private lateinit var currentPath: PointToPointPath
+    private lateinit var currentPaint: Paint
+    private lateinit var viewDimensions: Pair<Int, Int>
```

Comment:
There is a class `android.util.Size`, maybe it's a better fit than `Pair`?
(It doesn't support destructuring out of the box, but you can add the corresponding extension methods `component1()` and `component2()` to fix that, I believe)